perf/uncore_libpfm.go

```diff
@@ -0,0 +1,390 @@
+	if err != nil {
+		return nil, err
+	}
+	typeString := strings.TrimSpace(string(buf))
+	eventType, err := strconv.ParseUint(typeString, 0, 32)
+	if err != nil {
+		return nil, err
+	}
+
+	buf, err = ioutil.ReadFile(filepath.Join(path, pmuCpumaskFilename))
+	if err != nil {
+		return nil, err
+	}
+	var cpus []uint32
+	cpumask := strings.TrimSpace(string(buf))
+	for _, cpu := range cpumaskRegexp.Split(cpumask, -1) {
+		parsedCPU, err := strconv.ParseUint(cpu, 0, 32)
+		if err != nil {
+			return nil, err
+		}
+		cpus = append(cpus, uint32(parsedCPU))
+	}
+
+	return &pmu{name: name, typeOf: uint32(eventType), cpus: cpus}, nil
+}
+
+func getUncorePMUs(devicesPath string) (uncorePMUs, error) {
+	pmus := make(uncorePMUs, 0)
+
+	// Depends on platform, cpu mask could be for example in form "0-1" or "0,1".
+	cpumaskRegexp := regexp.MustCompile("[-,\n]")
+	err := filepath.Walk(devicesPath, func(path string, info os.FileInfo, err error) error {
+		// Skip root path.
+		if path == devicesPath {
+			return nil
+		}
+		if info.IsDir() {
+			if strings.Contains(info.Name(), uncorePMUPrefix) {
+				pmu, err := readUncorePMU(path, info.Name(), cpumaskRegexp)
+				if err != nil {
+					return err
+				}
+				pmus[info.Name()] = *pmu
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pmus, nil
+}
+
+type uncoreCollector struct {
+	cpuFiles           map[string]map[string]map[int]readerCloser
+	cpuFilesLock       sync.Mutex
+	events             [][]Event
+	eventToCustomEvent map[Event]*CustomEvent
+
+	// Handle for mocking purposes.
+	perfEventOpen func(attr *unix.PerfEventAttr, pid int, cpu int, groupFd int, flags int) (fd int, err error)
+}
+
+func NewUncoreCollector(cgroupPath string, events Events) stats.Collector {
+
+	if cgroupPath != rootPerfEventPath {
+		// Uncore metric doesn't exists for cgroups, only for entire platform.
+		return &stats.NoopCollector{}
+	}
+
+	collector := &uncoreCollector{}
+
+	// Default implementation of Linux perf_event_open function.
+	collector.perfEventOpen = unix.PerfEventOpen
+
+	err := collector.setup(events, systemDevicesPath)
+	if err != nil {
+		formatedError := fmt.Errorf("unable to setup uncore perf event collector: %v", err)
+		klog.V(5).Infof("Perf uncore metrics will not be available: %s", formatedError)
+		panic("elo")
+		return &stats.NoopCollector{}
+	}
+
+	return collector
+}
+
+func (c *uncoreCollector) setup(events Events, devicesPath string) error {
+	var err error
+	readUncorePMUs, err := getUncorePMUs(devicesPath)
+	if err != nil {
+		return err
+	}
+
+	// Maping from event name, pmu type, cpu.
+	c.cpuFiles = make(map[string]map[string]map[int]readerCloser)
+	c.events = events.UncoreEvents
+	c.eventToCustomEvent = parseUncoreEvents(events)
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+
+	for _, group := range c.events {
+		if len(group) > 1 {
+			klog.Warning("grouping uncore perf events is not supported!")
```

Comment:
is the plan to support this eventually, or is grouping uncore events never something we can support?

Reply:
Yes, after the merge I want to work on PR with grouping events.